### PR TITLE
add sparse regression example

### PR DESCRIPTION
**Pull request by martinjankowiak**
we port the sparse regression [example](https://github.com/pyro-ppl/numpyro/blob/master/examples/sparse_regression.py) from numpyro to pyro. in contrast to numpyro, where we do HMC inference for the kernel hyperparameters, here we do a MAP estimate. this works better than you might expect. along the way, we show how to use autodeltas to do MAP.

### CODE REVIEW

File: examples/sparse_regression.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+            init_losses.append(loss_fn(model, guide, X, Y, hypers).item())
+
+    pyro.set_rng_seed(np.argmin(init_losses))
+    pyro.clear_param_store()
+    guide = AutoDelta(model, init_loc_fn=functools.partial(init_to_median, num_samples=50))
+    with torch.no_grad():
+        loss_fn(model, guide, X, Y, hypers)
+    # we also make sure the initial observation noise is not too large
+    # (otherwise we run the danger of getting stuck in bad local optima during optimization).
+    pyro.param('auto_var_obs').unconstrained().data -= 2.0
```
Comment (jpchen) [inline examples/sparse_regression.py]: is 2.0 empirically selected?

File: examples/sparse_regression.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+    report_frequency = 50
+    print("Beginning MAP optimization...")
+
+    # the optimization loop
+    for step in range(args.num_steps):
+        loss = loss_fn(model, guide, X, Y, hypers) / args.num_data
+        loss.backward()
+        adam.step()
+        adam.zero_grad()
```
Comment (jpchen) [inline examples/sparse_regression.py]: this should happen before `step()`

File: examples/sparse_regression.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+        # we manually reduce the learning rate according to this schedule
+        if step in [100, 300, 700, 900]:
+            adam.param_groups[0]['lr'] *= 0.2
```
Comment (jpchen) [inline examples/sparse_regression.py]: there's a [scheduler](https://pytorch.org/docs/stable/optim.html?highlight=scheduler#torch.optim.lr_scheduler.MultiStepLR) for this

File: examples/sparse_regression.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+import functools
+
+import numpy as np
+import torch
+import math
+
+import pyro
+import pyro.distributions as dist
+from pyro import poutine
+from pyro import sample, param
```
Comment (jpchen) [inline examples/sparse_regression.py]: nit: i know this is ported from numpyro but it'd be nice to use `pyro.sample()` as in all the other tutorials

Comment (martinjankowiak) [reply examples/sparse_regression.py]: ok

File: examples/sparse_regression.py (hunk 1 above)
Comment (martinjankowiak) [reply examples/sparse_regression.py]: how else

File: examples/sparse_regression.py (hunk 2 above)
Comment (martinjankowiak) [reply examples/sparse_regression.py]: the two are equivalent

File: examples/sparse_regression.py (hunk 3 above)
Comment (martinjankowiak) [reply examples/sparse_regression.py]: I wanted to control everything myself and show that's done

**Comment by martinjankowiak** [comment]
I originally had code support but removed it because this is already 300 linez

**Comment by martinjankowiak** [comment]
I originally had code support but removed it because this is already 300 lines

File: examples/sparse_regression.py (hunk 2 above)
Comment (jpchen) [reply examples/sparse_regression.py]: hm yeah, in 0.3 or something the tensor `grad`s werent by default initialized to zero so i've always followed the pattern in their tutorials.

**Comment by fehiepsi** [comment]
> because this is already 300 lines

:)) np, I'll review this PR today.

File: examples/sparse_regression.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+Compare to theorem 5.1 in reference [1].
+"""
+@torch.no_grad()
+def compute_posterior_stats(X, Y, msq, lam, eta1, xisq, c, var_obs, jitter=1.0e-4):
+    N, P = X.shape
+
+    # prepare for computation of posterior statistics for singleton weights
+    probe = torch.zeros((2, P, P), dtype=X.dtype, device=X.device)
+    probe[0, ...] = torch.eye(P, dtype=X.dtype, device=X.device)
+    probe[1, ...] = -torch.eye(P, dtype=X.dtype, device=X.device)
```
Comment (fehiepsi) [inline examples/sparse_regression.py]: +1 another cool observation! It would be nicer if we can make `probe` have shape `(P, 2, P)` to match the pairwise part (it is also easier to think that we are create P probes of shapes `(2, P)`).

File: examples/sparse_regression.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+
+    # generate artificial dataset
+    X, Y, expected_thetas, expected_quad_dims = get_data(N=args.num_data, P=P, S=S,
+                                                         Q=Q, sigma_obs=args.sigma)
+
+    loss_fn = Trace_ELBO().differentiable_loss
+
+    # We initialize the AutoDelta guide (for MAP estimation) with args.num_trials many
+    # initial parameters sampled from the vicinity of the median of the prior distribution
+    # and then continue optimizing with the best performing initialization.
```
Comment (fehiepsi) [inline examples/sparse_regression.py]: +1

File: examples/sparse_regression.py (hunk 5 above)
Comment (martinjankowiak) [reply examples/sparse_regression.py]: done
